### PR TITLE
[SDK-1335][SDK-1052][SDK-1355] Animations on Camera Render, and support for fly to

### PR DESCRIPTION
**Pull request by MadisonEhlers-Vertex**
## What

Changes to support animations on render.

## Ticket

https://vertexvis.atlassian.net/browse/SDK-1335
https://vertexvis.atlassian.net/browse/SDK-1052

## Test Plan

See test plan on FSS pr for comprehensive test plan:
https://github.com/Vertexvis/frame-streaming-service/pull/143

Can be tested in the viewer by adding a duration in ms in the camera render func:
```
      await scene.camera().update(updateCamera).render({
               animation: {
                   milliseconds: 500
               }
          });
```

I also added the support for the fly to operation that supports either a bounding box, item id, suppliedId or camera.

```
      camera
        .flyTo(q => q.withSuppliedId('suppliedId'))
        .render({
          animation: {
            milliseconds: 500,
          },
        });
```

## Dependencies

https://github.com/Vertexvis/frame-streaming-service/pull/143

### CODE REVIEW

File: packages/viewer/src/rendering/remote.ts [hunk 1]
```diff
@@ -20,6 +22,15 @@ export interface FrameResponse {
 
 export type RemoteRenderer = FrameRenderer<FrameRequest, FrameResponse>;
 
+export const easingMap = {
+  linear: vertexvis.protobuf.stream.EasingType.EASING_TYPE_LINEAR,
+  'ease-out-cubic': vertexvis.protobuf.stream.EasingType.EASING_TYPE_LINEAR,
+  'ease-out-quad': vertexvis.protobuf.stream.EasingType.EASING_TYPE_LINEAR,
+  'ease-out-quart': vertexvis.protobuf.stream.EasingType.EASING_TYPE_LINEAR,
+  'ease-out-sine': vertexvis.protobuf.stream.EasingType.EASING_TYPE_LINEAR,
+  'ease-out-expo': vertexvis.protobuf.stream.EasingType.EASING_TYPE_LINEAR,
```
Comment (danconner-vertex) [inline packages/viewer/src/rendering/remote.ts:31]: is it intentional for all of these to be set to `vertexvis.protobuf.stream.EasingType.EASING_TYPE_LINEAR`?

Comment (MadisonEhlers-Vertex) [reply packages/viewer/src/rendering/remote.ts:31]: Nope. Haha. Will fix

File: packages/viewer/src/rendering/remote.ts [hunk 2]
```diff
@@ -51,10 +67,27 @@ function requestFrame(api: StreamApi): RemoteRenderer {
   return req => {
     const corrId = req.correlationId || UUID.create();
     const timeout = req.timeoutInMs || DEFAULT_TIMEOUT_IN_MS;
+    const animation =
+      req.animation && req.animation.milliseconds
```
Comment (danschultz) [inline packages/viewer/src/rendering/remote.ts:71]: `req.animation?.milliseconds`

File: packages/viewer/src/rendering/remote.ts [hunk 3]
```diff
@@ -20,6 +22,20 @@ export interface FrameResponse {
...
+    vertexvis.protobuf.stream.EasingType.EASING_TYPE_EASE_OUT_CUBIC,
+  'ease-out-quad':
+    vertexvis.protobuf.stream.EasingType.EASING_TYPE_EASE_OUT_QUAD,
+  'ease-out-quart':
+    vertexvis.protobuf.stream.EasingType.EASING_TYPE_EASE_OUT_QUART,
+  'ease-out-sine':
+    vertexvis.protobuf.stream.EasingType.EASING_TYPE_EASE_OUT_SINE,
+  'ease-out-expo':
+    vertexvis.protobuf.stream.EasingType.EASING_TYPE_EASE_OUT_EXPO,
+};
```
Comment (danschultz) [inline packages/viewer/src/rendering/remote.ts:37]: This doesn't need to be exported.

File: packages/viewer/src/rendering/remote.ts [hunk 4]
```diff
@@ -51,10 +67,27 @@ function requestFrame(api: StreamApi): RemoteRenderer {
   return req => {
     const corrId = req.correlationId || UUID.create();
     const timeout = req.timeoutInMs || DEFAULT_TIMEOUT_IN_MS;
+    const animation =
+      req.animation && req.animation.milliseconds
+        ? {
+            duration: {
+              nanos: (req.animation.milliseconds % 1000) * 1000000,
+              seconds: req.animation.milliseconds / 1000,
+            },
```
Comment (danschultz) [inline packages/viewer/src/rendering/remote.ts:76]: You can use `toProtoDuration` from the `stream-api` package.

File: packages/viewer/src/scenes/camera.ts [hunk 5]
```diff
@@ -67,9 +67,9 @@ export class Camera implements FrameCamera.FrameCamera {
    * Queues the rendering for a new frame using this camera. The returned
    * promise will resolve when a frame is received that contains this camera.
    */
-  public async render(): Promise<void> {
+  public async render(animation?: Animation.Animation): Promise<void> {
```
Comment (danschultz) [inline packages/viewer/src/scenes/camera.ts:70]: I'd maybe rather have an object like this instead:

```ts
interface CameraRenderOptions {
  animation: Animation.Animation;
}
```

This'd more easily allow us to add additional fields without possible breaking changes.

File: packages/viewer/src/types/animation.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
+export interface Animation {
+  milliseconds?: number;
+  easing?:
+    | 'linear'
+    | 'ease-out-cubic'
+    | 'ease-out-quad'
+    | 'ease-out-quart'
+    | 'ease-out-sine'
+    | 'ease-out-expo';
```
Comment (danschultz) [inline packages/viewer/src/types/animation.ts:9]: IMO `milliseconds` should be a required fields, and could be convinced that `easing` is as well.

File: packages/viewer/src/rendering/remote.ts [hunk 7]
```diff
@@ -51,10 +67,27 @@ function requestFrame(api: StreamApi): RemoteRenderer {
...
+            easing: req.animation?.easing
+              ? easingMap[req.animation.easing]
+              : undefined,
```
Comment (danschultz) [inline packages/viewer/src/rendering/remote.ts:79]: What happens if easing is undefined?

Comment (MadisonEhlers-Vertex) [reply packages/viewer/src/rendering/remote.ts:79]: I used the same behavior in ISS and it defaults to a linear interpolation.

Comment (MadisonEhlers-Vertex) [reply packages/viewer/src/rendering/remote.ts]: Addressed by making easing a required property on the animation.

**Review by danschultz — APPROVED** [review]

**Comment by MadisonEhlers-Vertex** [comment]
Going to hold off merging. With the new slerp algorithm, it makes sense to not have easing functions. Also we will be using a specific fly to part later so these changes can wait until those backend pieces are done.

File: packages/viewer/src/scenes/camera.ts [hunk 8]
```diff
@@ -50,6 +110,15 @@ export class Camera implements FrameCamera.FrameCamera {
     return this.update({ lookAt, position });
   }
 
+  /**
+   *
```
Comment (jdm717) [inline packages/viewer/src/scenes/camera.ts:114]: Might be worth adding a description/example to this method similar to what we have for the `scene.items` method

**Review by jdm717 — APPROVED** [review]
LGTM, just had one comment on documenting `flyTo`

**Review by jdm717 — APPROVED** [review]

**Pull request merged**